File: java/selenium3/src/test/java/com/titusfortner/selenium3/ShadowDomTest.java
package com.titusfortner.selenium3;

import io.github.bonigarcia.wdm.WebDriverManager;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.openqa.selenium.remote.RemoteWebDriver;
import org.openqa.selenium.remote.RemoteWebElement;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
public class ShadowDomTest {

    WebDriver driver;
    private WebElement shadowRoot;

    // What most people use for Shadow DOM Elements;
    // Still works for Chrome < v96, Edge < v96, Safari
    @Test
    public void oldCodeOldChrome() throws MalformedURLException {
        DesiredCapabilities caps = new DesiredCapabilities();
        caps.setBrowserName("chrome");
        caps.setCapability("browserVersion", "95.0");

        Map<String, Object> sauceOptions = new HashMap<>();
        sauceOptions.put("username", System.getenv("SAUCE_USERNAME"));
        sauceOptions.put("accessKey", System.getenv("SAUCE_ACCESS_KEY"));

        caps.setCapability("sauce:options", sauceOptions);
        String sauceUrl = "https://ondemand.saucelabs.com/wd/hub/";

        WebDriver driver = new RemoteWebDriver(new URL(sauceUrl), caps);

        driver.get("http://watir.com/examples/shadow_dom.html");

        WebElement shadowHost = driver.findElement(By.cssSelector("#shadow_host"));
        JavascriptExecutor jsDriver = (JavascriptExecutor) driver;

        WebElement shadowRoot = (WebElement) jsDriver.executeScript("return arguments[0].shadowRoot", shadowHost);
        WebElement shadowContent = shadowRoot.findElement(By.cssSelector("#shadow_content"));

        Assertions.assertEquals("some text", shadowContent.getText());
    }

    // Same code as above, but in Chromium 96+ it throws ClassCastException
    @Test
    public void oldCodeBroken() {
        WebDriverManager.chromedriver().setup();
        driver = new ChromeDriver();
        driver.get("http://watir.com/examples/shadow_dom.html");

        WebElement shadowHost = driver.findElement(By.cssSelector("#shadow_host"));
        JavascriptExecutor jsDriver = (JavascriptExecutor) driver;

        Object shadowRootObject = jsDriver.executeScript("return arguments[0].shadowRoot", shadowHost);
        Assertions.assertThrows(ClassCastException.class, () -> shadowRoot = (WebElement) shadowRootObject);

    }

    // If you absolutely *have to use Shadow Dom Elements in latest Chromium versions
    // in Selenium 3, this is the way to do it. It is hacky, so please update to Selenium 4 as soon as you can
    @Test
    public void hackWorks() {
        // Requires Chrome v96+, Edge v96+
        WebDriverManager.chromedriver().setup();
        driver = new ChromeDriver();

        driver.get("http://watir.com/examples/shadow_dom.html");

        WebElement shadow_host = driver.findElement(By.cssSelector("#shadow_host"));

        Object shadowRoot = ((JavascriptExecutor) driver).executeScript("return arguments[0].shadowRoot", shadow_host);
        String id = (String) ((Map<String, Object>) shadowRoot).get("shadow-6066-11e4-a52e-4f735466cecf");
        RemoteWebElement shadowRootElement = new RemoteWebElement();
        shadowRootElement.setParent((RemoteWebDriver) driver);
        shadowRootElement.setId(id);

        WebElement shadowContent = shadowRootElement.findElement(By.cssSelector("#shadow_content"));

        Assertions.assertEquals("some text", shadowContent.getText());
    }

    // Firefox is special
    @Test
    public void firefoxWorkaround() {
        WebDriverManager.firefoxdriver().setup();
        driver = new FirefoxDriver();
        driver.get("http://watir.com/examples/shadow_dom.html");

        WebElement shadowHost = driver.findElement(By.cssSelector("#shadow_host"));
        JavascriptExecutor jsDriver = (JavascriptExecutor) driver;

        List<WebElement> children = (List<WebElement>) jsDriver.executeScript("return arguments[0].shadowRoot.children", shadowHost);

        WebElement shadowContent = null;
        for (WebElement element : children) {
            if (element.getAttribute("id").equals("shadow_content")) {
                shadowContent = element;
                break;
            }
        }

        assert shadowContent != null;
        Assertions.assertEquals("some text", shadowContent.getText());
    }

    @AfterEach
    public void quit() {
        if (driver != null) {
            driver.quit();
        }
    }
}

File: dotnet/Selenium3/ShadowDomTest.cs
using System;
using System.Collections.Generic;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Firefox;
using OpenQA.Selenium.Remote;
using WebDriverManager;
using WebDriverManager.DriverConfigs.Impl;

namespace Selenium3
{
    [TestClass]
    public class ShadowDom
    {
        private IWebDriver _driver;

        [TestCleanup]
        public void CleanUpAfterEveryTestMethod()
        {
            if (_driver == null) return;
            _driver.Quit();
        }

        private void StartSauceDriver(string browserVersion)
        {
            var username = Environment.GetEnvironmentVariable("SAUCE_USERNAME");
            var accessKey = Environment.GetEnvironmentVariable("SAUCE_ACCESS_KEY");

            var browserOptions = new ChromeOptions()
            {
                BrowserVersion = browserVersion,
                PlatformName = "Windows 10"
            };

            var addr = "https://" + username + ":" + accessKey + "@ondemand.saucelabs.com/wd/hub";
            
            _driver = new RemoteWebDriver(new Uri(addr), 
                browserOptions.ToCapabilities());
        }

        // What most people use for Shadow DOM Elements;
        // Still works for Chrome < v96, Edge < v96, Safari
        [TestMethod]
        public void OldCodeOldChrome()
        {
            StartSauceDriver("95.0");

            _driver.Navigate().GoToUrl("http://watir.com/examples/shadow_dom.html");

            var shadowHost = _driver.FindElement(By.CssSelector("#shadow_host"));
            var js = ((IJavaScriptExecutor)_driver);

            var shadowRoot = (IWebElement)js.ExecuteScript("return arguments[0].shadowRoot", shadowHost);
            var shadowContent = shadowRoot.FindElement(By.CssSelector("#shadow_content"));

            Assert.AreEqual("some text", shadowContent.Text);
        }
     
        // Same code as above, but in Chromium 96+ it throws InvalidCastException
        [TestMethod]
        [ExpectedException(typeof(InvalidCastException),"No Longer Works with IWebElement")]
        public void OldCodeBroken()
        {
            new DriverManager().SetUpDriver(new ChromeConfig());
            _driver = new ChromeDriver();

            _driver.Navigate().GoToUrl("http://watir.com/examples/shadow_dom.html");

            var shadowHost = _driver.FindElement(By.CssSelector("#shadow_host"));
            var js = ((IJavaScriptExecutor)_driver);

            var shadowRoot = (IWebElement)js.ExecuteScript("return arguments[0].shadowRoot", shadowHost);
        }
        
        // If you absolutely *have to use Shadow Dom Elements in latest Chromium versions
        // in Selenium 3, this is the way to do it. It is hacky, so please update to Selenium 4 as soon as you can
        [TestMethod]
        public void HackWorks()
        {
            new DriverManager().SetUpDriver(new ChromeConfig());
            var options = new ChromeOptions();
            _driver = new ChromeDriver(options);

            _driver.Navigate().GoToUrl("http://watir.com/examples/shadow_dom.html");

            var shadowHost = _driver.FindElement(By.CssSelector("#shadow_host"));
            var js = ((IJavaScriptExecutor)_driver);

            var shadowRoot = (Dictionary<string, object>)js.ExecuteScript("return arguments[0].shadowRoot", shadowHost);
            var id = (string)shadowRoot["shadow-6066-11e4-a52e-4f735466cecf"];
            var shadowRootElement = new RemoteWebElement((RemoteWebDriver)_driver, id);

            var shadowContent = shadowRootElement.FindElement(By.CssSelector("#shadow_content"));

            var text = shadowContent.Text;

            Assert.AreEqual("some text", text);
        }

        // Firefox is special
        [TestMethod]
        public void FirefoxWorkaround()
        {
            new DriverManager().SetUpDriver(new FirefoxConfig());
            _driver = new FirefoxDriver();

            _driver.Navigate().GoToUrl("http://watir.com/examples/shadow_dom.html");

            var shadowHost = _driver.FindElement(By.CssSelector("#shadow_host"));
            var js = ((IJavaScriptExecutor)_driver);

            var children = (IEnumerable<IWebElement>)js.ExecuteScript("return arguments[0].shadowRoot.children", shadowHost);

            IWebElement shadowContent = null;
            foreach (IWebElement element in children) {
                if (element.GetAttribute("id").Equals("shadow_content")) {
                    shadowContent = element;
                    break;
                }
            }

            Assert.AreEqual("some text", shadowContent?.Text);
        }
    }
}
